### PR TITLE
Support binding a graph output to a specific device via the Python binding

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: This change supports binding a graph output to a specific device (like GPU) without the requiring a pre-allocated output buffer to do so (ORT will internally allocate and manage the buffer). For this release, the user won't have access to the ORT allocated buffer and hence won't directly be able to access its contents (we don't support an equivalent of an OrtValue in the Python world - we need to design of the life cycle management for such a struct and we will probably do this for the next release). The user can call the exposed get_outputs() which will copy the contents to the host and return back the result as a Numpy object. This provides the advantage that Run() will not incur the device-host copying cost (as is the case today) and helps the benchmarking scenario. This feature is in addition to the already exposed feature where-in a user can use a custom allocator (for example - Torch) to pre-allocate memory on a device and have ORT directly write results into the buffer.

Brings in this bug fix as it is required - https://github.com/microsoft/onnxruntime/pull/4415 

**Motivation and Context**
Support IOBinding via Python API

### CODE REVIEW

File: onnxruntime/core/session/IOBinding.cc [hunk 1]
```diff
@@ -78,7 +78,16 @@ common::Status IOBinding::SynchronizeOutputs() {
   return Status::OK();
 }
 
-common::Status IOBinding::BindOutput(const std::string& name, const OrtValue& ml_value, OrtDevice device) {
+common::Status IOBinding::BindOutput(const std::string& name, const OrtValue& ml_value) {
```
Comment (hariharans29) [inline onnxruntime/core/session/IOBinding.cc:81]: minor refactoring as suggested by @skottmckay  - split BindOutput() into 2 - one with an OrtValue (pre-allocated memory) and one with OrtDevice (no pre-allocation + telling Ort where to allocate). Removes all confusions with having both parameters together. 

CC: @yuslepukhin as this may be important info when working on C/C# APIs for IOBinding

File: onnxruntime/core/session/inference_session.cc [hunk 2]
```diff
@@ -995,6 +995,10 @@ const SessionOptions& InferenceSession::GetSessionOptions() const {
   return session_options_;
 }
 
+const DataTransferManager& InferenceSession::GetDataTransferManager() const {
+  return data_transfer_mgr_;
```
Comment (hariharans29) [inline onnxruntime/core/session/inference_session.cc:999]: Need DataTransferManager in pybind to do the device-host copy if the tensor is associated with memory in CUDA but the user requests a Numpy object

File: onnxruntime/python/session.py [hunk 3]
```diff
@@ -228,9 +228,19 @@ def bind_output(self, name, device_type='cpu', device_id=0, element_type=None, s
...
+        # would like to bind an output to a specific device (e.g. cuda).
+        # Pre-allocating an output buffer may not be an option for the user as :
+        # (1) They may not want to use a custom allocator specific to the device they want to bind the output to,
+        # in which case ORT will allocate the memory for the user 
+        # (2) The output has a dynamic shape and hence the size of the buffer may not be fixed across runs
+        if buffer_ptr is None:
+            self._iobinding.bind_output(name, C.OrtDevice(get_ort_device_type(device_type), C.OrtDevice.default_memory(),
+                                                          device_id))
         else:
+            if element_type is None or shape is None:
```
Comment (hariharans29) [inline onnxruntime/python/session.py:242]: raise error if missing metadata associated with pre-allocated memory

File: onnxruntime/test/python/onnxruntime_test_python_iobinding.py [hunk 4]
```diff
@@ -42,31 +43,94 @@ def create_model_and_input(self):
...
+
+        session = onnxruntime.InferenceSession('model.onnx')
+        io_binding = session.io_binding()
+
+        # Bind input to CUDA
+        io_binding.bind_input('input', torch_input.device.type, 0, np.float32, list(torch_input.size()), torch_input.data_ptr())
+
+        # Bind output to CUDA
+        io_binding.bind_output('output', 'cuda')
+
```
Comment (skottmckay) [inline onnxruntime/test/python/onnxruntime_test_python_iobinding.py:123]: nit: Maybe add a comment saying this doesn't really buy you anything at this point as the only current way to access the output is via the copy to CPU in io_binding.get_outputs, and that this will be addressed in the future. Otherwise someone may read this script and get confused about what this achieves.

File: onnxruntime/test/python/onnxruntime_test_python_iobinding.py [hunk 5]
```diff
@@ -42,31 +43,94 @@ def create_model_and_input(self):
...
+        # Bind output to CUDA
+        ort_output = torch.empty(list(torch_output.size())).cuda()
+        io_binding.bind_output('output', torch_output.device.type, 0, np.float32, list(torch_output.size()), ort_output.data_ptr())
+
+        # Invoke Run
+        session.run_with_iobinding(io_binding)
+        
+        # Get outputs over to CPU (the outputs which were bound to CUDA will get copied over to the host here)
+        ort_output_vals = io_binding.get_outputs()[0]
```
Comment (hariharans29) [inline onnxruntime/test/python/onnxruntime_test_python_iobinding.py:107]: #4415 had o`rt_output_vals = ort_output.cpu().detach().numpy()` to copy over the memory contents from CUDA to CPU and return a Numpy object. Felt it is meaningful to use `get_outputs()` and tests its functionality of making the device-host copy.

File: tools/ci_build/build.py [hunk 6]
```diff
@@ -1200,6 +1200,22 @@ def run_onnxruntime_tests(args, source_dir, ctest_path, build_dir, configs):
                 [sys.executable, 'onnxruntime_test_python.py'],
                 cwd=cwd, dll_path=dll_path)
 
+            iobinding_test = False
```
Comment (hariharans29) [inline tools/ci_build/build.py:1203]: Make sure tests are run in atleast one build pipeline. This should now run in Linux GPU CI pipeline.

File: tools/ci_build/github/linux/docker/scripts/install_deps.sh [hunk 7]
```diff
@@ -113,8 +113,8 @@ if [ $DEVICE_TYPE = "Normal" ]; then
     ${PYTHON_EXE} -m pip install sympy==1.1.1
 elif [ $DEVICE_TYPE = "gpu" ]; then
     ${PYTHON_EXE} -m pip install sympy==1.1.1
+    ${PYTHON_EXE} -m pip install --upgrade --pre torch==1.6.0.dev20200610 torchvision==0.7.0.dev20200610 -f https://download.pytorch.org/whl/nightly/cu101/torch_nightly.html
```
Comment (hariharans29) [inline tools/ci_build/github/linux/docker/scripts/install_deps.sh:116]: Inferencing CI pipelines need Torch installed for testing IOBinding

File: onnxruntime/test/python/onnxruntime_test_python_iobinding.py (hunk 5 above)
Comment (hariharans29) [reply onnxruntime/test/python/onnxruntime_test_python_iobinding.py:107]: ort_output.cpu().detach().numpy() must be tested too as that is the only way to make sure ORT actually wrote into the pre-allocated buffer. Thanks @skottmckay

File: onnxruntime/test/python/onnxruntime_test_python_iobinding.py (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/test/python/onnxruntime_test_python_iobinding.py:123]: sure, I can add a comment

Comment (hariharans29) [reply onnxruntime/test/python/onnxruntime_test_python_iobinding.py:135]: added

**Comment by tianleiwu** [comment]
For life cycle management: will the output buffer be overwritten (and re-allocated) in next inference call? 

For GPT-2 case, the past state output will feed as input to next inference call. So the output is better to stay in GPU, and can be able to pass it as input. Could we design the interface to support such senario? We need not implement it right now, but it is better to have design in place.

**Comment by skottmckay** [comment]
We need to add something that essentially makes an OrtValue available in python. With that I assume we could update io_binding.get_outputs to return instances of that type, and you could bind those instances to inputs for the next inference call. 

---
In reply to: [654582048](https://github.com/microsoft/onnxruntime/pull/4439#issuecomment-654582048) [](ancestors = 654582048)

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 8]
```diff
@@ -460,13 +460,13 @@ void addGlobalMethods(py::module& m, const Environment& env) {
             onnxruntime::CreateExecutionProviderFactory_MIGraphX(0)
 #endif
 #ifdef USE_VITISAI
-            onnxruntime::CreateExecutionProviderFactory_VitisAI("DPU", 0),
+                onnxruntime::CreateExecutionProviderFactory_VitisAI("DPU", 0),
```
Comment (ytaous) [inline onnxruntime/python/onnxruntime_pybind_state.cc:463]: nit : format, same below

**Comment by hariharans29** [comment]
> We need to add something that essentially makes an OrtValue available in python. With that I assume we could update io_binding.get_outputs to return instances of that type, and you could bind those instances to inputs for the next inference call.
> 
> In reply to: [654582048](https://github.com/microsoft/onnxruntime/pull/4439#issuecomment-654582048) [](ancestors = 654582048)

Ideally, yes. get_outputs() will have a revised signature to return instances of OrtPythonValue (proxy) but I don’t know if we can modify the signatures if released APIs given that it would be a breaking change.

**Comment by hariharans29** [comment]
> For life cycle management: will the output buffer be overwritten (and re-allocated) in next inference call?
> 
> For GPT-2 case, the past state output will feed as input to next inference call. So the output is better to stay in GPU and can pass as input. Could we design the interface to support such scenario? We need not implement it right now, but it is better to have design in place.

I think this should be possible. The output will stay in GPU and be re-used for the next run unless the user has destroyed it in which case it won’t be available.

**Comment by hariharans29** [comment]
> > We need to add something that essentially makes an OrtValue available in python. With that I assume we could update io_binding.get_outputs to return instances of that type, and you could bind those instances to inputs for the next inference call.
> > In reply to: [654582048](https://github.com/microsoft/onnxruntime/pull/4439#issuecomment-654582048) [](ancestors = 654582048)
> 
> Ideally, yes. get_outputs() will have a revised signature to return instances of OrtPythonValue (proxy) but I don’t know if we can modify the signatures if released APIs given that it would be a breaking change.

Thinking along these lines - should we just rename `get_outputs()` to `get_cpu_outputs()` (something similar to `bind_cpu_input()` for binding Numpy objects to inputs). `get_cpu_outputs()` would always return Numpy objects and we can reserve `get_outputs()` for the future call that will return `OrtPythonValue` instances (or whatever they are called). Thoughts - @pranavsharma, @skottmckay, @duli2012  ?

File: onnxruntime/python/onnxruntime_pybind_state.cc (hunk 8 above)
Comment (hariharans29) [reply onnxruntime/python/onnxruntime_pybind_state.cc:463]: This is the formatting I get after running clang-format

**Comment by skottmckay** [comment]
> Thinking along these lines - should we just rename `get_outputs()` to `get_cpu_outputs()` (something similar to `bind_cpu_input()` for binding Numpy objects to inputs). `get_cpu_outputs()` would always return Numpy objects and we can reserve `get_outputs()` for the future call that will return `OrtPythonValue` instances (or whatever they are called). Thoughts - @pranavsharma, @skottmckay, @duli2012 ?

I'm fine with reserving a more flexible `get_outputs` for future usage. `get_outputs_on_cpu()` may be clearer though. `get_cpu_outputs` could be interpreted as get the outputs that were bound to cpu.

**Comment by pranavsharma** [comment]
> > > We need to add something that essentially makes an OrtValue available in python. With that I assume we could update io_binding.get_outputs to return instances of that type, and you could bind those instances to inputs for the next inference call.
> > > In reply to: [654582048](https://github.com/microsoft/onnxruntime/pull/4439#issuecomment-654582048) [](ancestors = 654582048)
> > 
> > 
> > Ideally, yes. get_outputs() will have a revised signature to return instances of OrtPythonValue (proxy) but I don’t know if we can modify the signatures if released APIs given that it would be a breaking change.
> 
> Thinking along these lines - should we just rename `get_outputs()` to `get_cpu_outputs()` (something similar to `bind_cpu_input()` for binding Numpy objects to inputs). `get_cpu_outputs()` would always return Numpy objects and we can reserve `get_outputs()` for the future call that will return `OrtPythonValue` instances (or whatever they are called). Thoughts - @pranavsharma, @skottmckay, @duli2012 ?

Good idea. I would in fact make it more explicit by calling it copy_outputs_to_cpu(). And we can reserve get_outputs() for the cascading inferencing use case.

**Comment by pranavsharma** [comment]
LGTM 👍

File: onnxruntime/test/python/onnxruntime_test_python_iobinding.py [hunk 9]
```diff
@@ -42,31 +43,106 @@ def create_model_and_input(self):
...
         ort_output = io_binding.get_outputs()[0]
-    
-        self.assertTrue(np.array_equal(torch_output, ort_output))
+
+        # Validate results
+        self.assertTrue(np.array_equal(torch_output_vals, ort_output))
 
     def test_bind_input_to_cpu_arr(self):
-        torch_input, torch_output = self.create_model_and_input()
+        torch_input, torch_outpu, torch_output_vals = self.create_model_and_input()
```
Comment (tianleiwu) [inline onnxruntime/test/python/onnxruntime_test_python_iobinding.py:70]: torch_output

**Comment by duli2012** [comment]
> > > > We need to add something that essentially makes an OrtValue available in python. With that I assume we could update io_binding.get_outputs to return instances of that type, and you could bind those instances to inputs for the next inference call.
> > > > In reply to: [654582048](https://github.com/microsoft/onnxruntime/pull/4439#issuecomment-654582048) [](ancestors = 654582048)
> > > 
> > > 
> > > Ideally, yes. get_outputs() will have a revised signature to return instances of OrtPythonValue (proxy) but I don’t know if we can modify the signatures if released APIs given that it would be a breaking change.
> > 
> > 
> > Thinking along these lines - should we just rename `get_outputs()` to `get_cpu_outputs()` (something similar to `bind_cpu_input()` for binding Numpy objects to inputs). `get_cpu_outputs()` would always return Numpy objects and we can reserve `get_outputs()` for the future call that will return `OrtPythonValue` instances (or whatever they are called). Thoughts - @pranavsharma, @skottmckay, @duli2012 ?
> 
> Good idea. I would in fact make it more explicit by calling it copy_outputs_to_cpu(). And we can reserve get_outputs() for the cascading inferencing use case.



> cascading

Second this.
